File: Test_Code/Classes/Field.py
from concurrent.futures import ThreadPoolExecutor
import multiprocessing
import numpy as np

##
# @file Field.py
# @brief This file contains the Field class, which is used to represent a field with a grid of cells.
# @version v1a
# @date 15.02.2025
# @package Field
# @brief Class for representing a field with a grid of cells
#

class Field:
    
    """!
    Field class.
    
    @brief Class for representing a field with a grid of cells.
    """
    def __init__(self, length_m, width_m, rows, cols, rx, ry, alpha, beta, Wm, Deltat, line = None, field=None):
        
        """!
        
        Constructor for Field class.
        
        @param length_m Length of the field in meters.
        @param width_m Width of the field in meters.
        @param rows Number of rows in the field.
        @param cols Number of columns in the field.
        @param rx Radius of watering in meters.
        @param ry Radius of watering in meters.
        @param alpha Coefficient for watering.
        @param beta Coefficient for watering.
        @param Wm Maximum watering in meters.
        @param Deltat Time step.
        @param line Line index of watering machine.
        @param field 2D np.array, optional
        @details Initializes the field with the given parameters and sets the moisture level of each cell to a random value between 0 and 1.
        
        """
        
        # Initializing the field
        self.rows = rows
        self.cols = cols
        self.rx = rx
        self.ry = ry
        self.alpha = alpha
        self.beta = beta
        self.Wm = Wm
        self.Deltat = Deltat
        self.line = line if line is not None else self.rows // 2  # Линия расположения машины
        self.field = field if field is not None else self.randomize_field()
        self.length_m = length_m
        self.width_m = width_m
        self.cell_length_m = length_m / cols  
        self.cell_width_m = width_m / rows  
        self.rx_cells = int(self.rx / self.cell_length_m)
        self.ry_cells = int(self.ry / self.cell_width_m)


    
    def calc_base(self, func, a = 2, b = 0.3, c = 3):
        """!
        Calculate the base value of the field.

        @param func Function to calculate the base value.
        @param a Parameter a.
        @param b Parameter b.
        @param c Parameter c.
        """
        
        return func(self, a=a, b=b, c=c)


    def __getitem__(self, index):
        if isinstance(index, int):
            return self.field[index]
        row, col = index
        return self.field[row][col]
    
    def __setitem__(self, index, value):
        if isinstance(index, int):
            self.field[index] = value
        else:
            row, col = index
            self.field[row][col] = value
    
    def __str__(self):
        base = f"Field with {self.rows} rows and {self.cols} columns\n" + f"Cell length: {self.cell_length_m} m, Cell width: {self.cell_width_m} m\n" + \
            f"Rx: {self.rx} m, Ry: {self.ry} m\n" + f"rx_cells: {self.rx_cells}, ry_cells: {self.ry_cells}\n"
        for row in self.field:
            base += str(row) + "\n"
        return base

    def randomize_field(self, MAX_V = 0.4123, MIN_V = 0.0):
        """!
        Randomize the field.

        @param MAX_V Maximum water level.
        @param MIN_V Minimum water level.
        """

        field = np.random.uniform(MIN_V, MAX_V, (self.rows, self.cols))
        field[self.line] = [-1 for _ in range(self.cols)]
        self.field = field
    
    
    def avg_field(self, field = None):
        """!
        
        Calculate the average water level in the field

        @param field 2D np.array, optional
        @details Calculates the average water level in the field.

        @return float
        """
        if field is None:
            field = self.field
        valid_cells = field[field != -1]  
        return np.mean(valid_cells)
    
    def update_cell(self, r, c, x, w, v):
        """!
        
        Update the moisture level of a cell in the field

        @param r Row index of the cell.
        @param c Column index of the cell.
        @param x Current position of watering machine.
        @param w Watering intensity.
        @param v Water level.

        @return float
        @details Updates the moisture level of a cell in the field based on the current position of the watering machine, the watering intensity, and the water level.
        
        """
        if self.field[r][c] == -1:
            return 0
        d_rc = ((r - self.line) ** 2 + (c - x) ** 2) ** 0.5
        term = w * (self.Wm * self.Deltat / (d_rc**2 + 1)**self.beta) * np.exp(-self.alpha * v)
        self.field[r][c] += term
        return term

    def parallel_update_field(self, x, w, v):
        """
        @param x Current position of watering machine.
        @param w Watering intensity.
        @param v Water level.
        @return List of changes in the field.
        @details Updates the moisture level of cells in the field based on the current position of the watering machine, the watering intensity, and the water level.
        
        """
        start_col = max(0, x - self.rx_cells)
        end_col = min(self.cols, x + 1)
        if start_col == end_col:
            start_col -= self.rx_cells

        max_workers = multiprocessing.cpu_count()
        
        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            futures = [
                executor.submit(self.update_cell, r, c, x, w, v)
                for r in range(max(0, self.line - self.ry_cells), min(self.rows, self.line + self.ry_cells + 1))
                for c in range(start_col, end_col)
            ]
            results = [future.result() for future in futures]

        return results  # Возвращает список изменений для анализа

File: cppCode/DIPLIM_console/DIPLIM_console/Field.cpp
#include "Field.h"
#include "ModelParams.h"
#include <cmath>
#include <thread>
#include <future>
#include <algorithm>
#include "Func.h"


Field::Field(ModelParams& Mparams)
{
	Field::FieldParams = Mparams;
	Field::FieldMap = Mparams.getModelParams("F");

	
	double cell_length_m = std::any_cast<double>(Field::FieldMap["length_m"]) / std::any_cast<int>(Field::FieldMap["cols"]);
	double cell_width_m = std::any_cast<double>(Field::FieldMap["width_m"]) / std::any_cast<int>(Field::FieldMap["rows"]);
	
	
	double rx_cells = (int)(std::any_cast<double>(Field::FieldMap["rx"]) / cell_length_m);
	double ry_cells = (int)(std::any_cast<double>(Field::FieldMap["ry"]) / cell_width_m);

	
	Field::FieldMap["rx_cells"] = rx_cells;
	Field::FieldMap["ry_cells"] = ry_cells;
	Field::FieldMap["cell_length_m"] = cell_length_m;
	Field::FieldMap["cell_width_m"] = cell_width_m;
	int line = std::any_cast<int>(Field::FieldMap["line"]);
	Field::FieldMap["line"] = line;
	Field::FieldMap2D = std::vector<std::vector<double>>(std::any_cast<int>(Field::FieldMap["rows"]), std::vector<double>(std::any_cast<int>(Field::FieldMap["cols"]), 0.0));
	FieldMap2D[line] = std::vector<double>(std::any_cast<int>(Field::FieldMap["cols"]), -1);
}

Field::~Field()
{
	Field::FieldMap.clear();
}

double Field::calc_base()
{
	double res = 0;
	int rows = std::any_cast<int>(Field::FieldMap["rows"]);
	int cols = std::any_cast<int>(Field::FieldMap["cols"]);
	double a = std::any_cast<double>(Field::FieldParams.getParam("M", "a"));
	double b = std::any_cast<double>(Field::FieldParams.getParam("M", "b"));
	double c = std::any_cast<double>(Field::FieldParams.getParam("M", "c"));
	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < cols; j++)
		{
			if ((*this)(i, j) == -1) continue;
			res += -a * pow(((*this)(i, j) - b), 2) + c;
		}
	}
	return res;
}

void Field::randomizeField(double min_value, double max_value)
{
	int rows = std::any_cast<int>(Field::FieldMap["rows"]);
	int cols = std::any_cast<int>(Field::FieldMap["cols"]);

	for ( int i = 0; i < rows; i++) {
		for (int j = 0; j < cols; j++) {
			if ((*this)(i, j) == -1) continue;
			double value = min_value + static_cast<double>(rand()) / (static_cast<double>(RAND_MAX) / (max_value - min_value));
			(*this)(i, j) = value;
		}
	}
}

double Field::avgerageField()
{
	int rows = std::any_cast<int>(Field::FieldMap["rows"]);
	int cols = std::any_cast<int>(Field::FieldMap["cols"]);
	double res = 0;
	for (int i = 0; i < rows; i++) {
		for (int j = 0; j < cols; j++) {
			double el = (*this)(i, j);
			if (el >= 0.0) res += el;
		}
	}
	return res / (rows * cols);
}

double Field::minField()
{
	double min = 0;
	int rows = std::any_cast<int>(Field::FieldMap["rows"]);
	int cols = std::any_cast<int>(Field::FieldMap["cols"]);
	for (int i = 0; i < rows; i++) {
		for (int j = 0; j < cols; j++) {
			double el = (*this)(i, j);
			if (el < min) min = el;
		}
	}
	return min;
}

double Field::maxField()
{
double max = 0;
	int rows = std::any_cast<int>(Field::FieldMap["rows"]);
	int cols = std::any_cast<int>(Field::FieldMap["cols"]);
	for (int i = 0; i < rows; i++) {
		for (int j = 0; j < cols; j++) {
			double el = (*this)(i, j);
			if (el > max) max = el;
		}
	}
	return max;
}

std::vector<double> Field::DataField()
{
	double avg = avgerageField();
	double min = minField();
	double max = maxField();
	std::vector<double> res = { avg, min, max };
	return res;
}

double Field::update_cell(int i, int j, int x, double w, double v)
{
	if ((*this)(i, j) == -1.0) return 0.0;
	double Wm = std::any_cast<double>(Field::FieldParams.getParam("M", "Wm"));
	double Dt = std::any_cast<double>(Field::FieldParams.getParam("M", "Deltat"));
	double beta = std::any_cast<double>(Field::FieldParams.getParam("M", "beta"));
	double alpha = std::any_cast<double>(Field::FieldParams.getParam("M", "alpha"));

	double d_ij = pow(pow(i - std::any_cast<int>(Field::FieldMap["line"]), 2) + pow((j - x), 2), 0.5);
	double term = w * (Wm * Dt / pow((pow(d_ij, 2) + 1), beta)) * exp(-alpha * v);

	(*this)(i, j) += term;

	return term;
}
std::vector<double> Field::update_field(int x, double w, double v)
{
	int rows = std::any_cast<int>(Field::FieldMap["rows"]);
	int cols = std::any_cast<int>(Field::FieldMap["cols"]);
	int rx_cells = std::any_cast<double>(Field::FieldMap["rx_cells"]);
	int ry_cells = std::any_cast<double>(Field::FieldMap["ry_cells"]);
	int line = std::any_cast<int>(Field::FieldMap["line"]);


	// start col, end col
	int start_row = std::max(0, line - ry_cells);
	int end_row = std::min(rows, line + ry_cells + 1);
	
	int start_col = MAX(0, x - rx_cells);
	int end_col = MIN(cols, x + rx_cells + 1);

	if (start_col == end_col) {
		start_col -= rx_cells;
	}

	std::vector<std::future<double>> futures;
	std::vector<double> results;

	// parallel update
	for (int i = start_row; i < end_row; i++) {
		for (int j = start_col; j < end_col; j++) {
			futures.push_back(std::async(std::launch::async, &Field::update_cell, this, i, j, x, w, v));
		}
	}

	// collect results
	for (auto& future : futures)
	{
		results.push_back(future.get());
	}
	
	return results;
}

std::ostream& operator<<(std::ostream& os, const Field& field)
{
	std::map<std::string, std::any> fieldMap = field.getFieldMap();

	// Преобразуем данные в строку с правильным форматированием
	std::string s = "Rows: " + std::to_string(std::any_cast<int>(fieldMap["rows"])) +
		" Columns: " + std::to_string(std::any_cast<int>(fieldMap["columns"])) + "\n";
	s += "Length: " + std::to_string(std::any_cast<double>(fieldMap["length_m"])) +
		" Width: " + std::to_string(std::any_cast<double>(fieldMap["width_m"])) + "\n";
	s += "Rx: " + std::to_string(std::any_cast<double>(fieldMap["rx"])) +
		" Ry: " + std::to_string(std::any_cast<double>(fieldMap["ry"])) +
		s += "Rx (cells): " + std::to_string(std::any_cast<double>(fieldMap["rx_cells"])) +
		" Ry (cells): " + std::to_string(std::any_cast<double>(fieldMap["ry_cells"])) + "\n";  // Исправлено Rz

	return os << s;
}

